feat: LEAP-1199: Add drag and drop functionality on DM tabs

juliosgarbi opened the pull request:
### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [x] Backend (API)
- [x] Frontend



### Describe the reason for change
Currently, users cannot move tabs around, which is inconvenient. They must delete ("Close") tabs and recreate them in the desired order. This update allows users to reorder tabs by dragging and dropping them.

Key features:

- Drag and drop functionality for tabs.
- Visual indicator to show the drop location.
- Dragging not confined to vertical tab limits.
- Tabs can be reordered regardless of visibility.
- Order of tabs is consistent across the organization after a hard refresh.
- No disruption for concurrent users.
- ESC key cancels the drag action, returning the tab to its original position.


![Screen-Recording-2024-07-16-at-18 09 18](https://github.com/user-attachments/assets/2e3a5896-bcd4-4d0c-b695-3d8b88b5899f)

juliosgarbi committed:
feat: LEAP-1199: Add drag and drop functionality on DM tabs

triklozoid commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/9957138605)
> Successfully merged:  rename web/dist/apps/labelstudio/{book.d787a03585e3dfbb093186a4f7e882d3.svg => book.23db5e3751679a820c6e975e99f29323.svg} (70%)

triklozoid committed:
Merge branch 'develop' into 'fb-leap-1199'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/9957138605

triklozoid committed:
feat: LEAP-1241: Views ordering (#6091)

* feat: LEAP-1241: Views ordering

* Fix urls test

* Add index for project+order field

---------

Co-authored-by: triklozoid <triklozoid@users.noreply.github.com>

bmartel committed:
fix: OPTIC-885: Links not clickable from within Radio Button (#6097)

* fix: OPTIC-885: Links not clickable from within Radio Button

* add a comment stating this fix needs to be revisited

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/9957459774

juliosgarbi committed:
Merge branch 'develop' into fb-leap-1199

juliosgarbi commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/9960418462)
> Successfully merged: Already up to date.

juliosgarbi committed:
add ordertab api

triklozoid committed:
Switch method to POST instead of PUT

juliosgarbi committed:
Merge branch 'fb-leap-1199' of https://github.com/HumanSignal/label-studio into fb-leap-1199

juliosgarbi committed:
set the order tab api

juliosgarbi committed:
fix the fetch API to be able to receive an empty document from a request

jombooth commented on label_studio/data_manager/models.py:
```diff
@@ -9,6 +9,7 @@
 class ViewBaseModel(models.Model):
     data = models.JSONField(_('data'), default=dict, null=True, help_text='Custom view data')
     ordering = models.JSONField(_('ordering'), default=dict, null=True, help_text='Ordering parameters')
+    order = models.IntegerField(_('order'), default=0, help_text='Order of view')
```
This migration isn't compatible with our migration guidelines - we shouldn't be adding new fields that aren't nullable, because they cause problems if we have to revert.

jombooth replied:
While we're here, I think the help text could clarify the purpose of this field more; "order of view" could mean a lot of different things. We should spell out that it's the position of the tab, starting at the left in data manager and increasing as the tabs go left to right.

triklozoid replied:
Fixed.

jombooth reviewed (changes requested):
We need to rework the migration that's included in this PR to make the new field nullable; code depending on the new field should handle nulls.

triklozoid committed:
Make order field nullable

hlomzik commented on web/libs/datamanager/src/components/Common/Tabs/Tabs.jsx:
```diff
@@ -7,12 +7,13 @@ import { Icon } from "../Icon/Icon";
 import Input from "../Input/Input";
 import "./Tabs.styl";
 import { TabsMenu } from "./TabsMenu";
+import { DragDropContext, Droppable } from "react-beautiful-dnd";
```
could you please move it up to be in order?

hlomzik commented on web/libs/datamanager/src/components/DataManager/DataManager.jsx:
```diff
@@ -10,6 +10,7 @@ import { FiltersSidebar } from "../Filters/FiltersSidebar/FilterSidebar";
 import { DataView } from "../MainView";
 import "./DataManager.styl";
 import { Toolbar } from "./Toolbar/Toolbar";
+import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
```
and here as well

hlomzik commented on web/libs/datamanager/src/stores/Tabs/store.js:
```diff
@@ -333,6 +333,23 @@ export const TabStore = types
+      const sn = getSnapshot(removed);
+
+      // Insert the view at the new position
+      self.views.splice(destination, 0, sn);
```
Probably these manipulations are the reason for errors on tab duplication.
`self.views` is changed in place, so no re-renders happen. So rendered tabs are the same old objects even after updates, because original tab is removed and new one is created from snapshot. So rendered tab doesn't have parent anymore.
To fix this you can first try to update `views` in immutable manner by updating it: `self.views = [...self.views]`. Then all rendered tabs will be the actual ones, but this can slow tabs down. You can also try to change the way you access parent, basically it's just `views`.

juliosgarbi replied:
Yeah, I started using views instead tab.parent, it's easier and it doesn't depend of any reference, I believe that in this case this solution is better. Thanks @hlomzik

juliosgarbi committed:
fix the tab duplication when user reorders the tab

hlomzik commented on web/libs/datamanager/src/components/Common/Tabs/Tabs.jsx:
```diff
@@ -37,8 +47,16 @@ export const Tabs = ({ children, activeTab, onChange, onAdd, tabBarExtraContent,
+          <DragDropContext onDragEnd={onDragEnd}>
+            <Droppable droppableId="droppable" direction="horizontal">
+              {(provided) => (
+                <Elem ref={provided.innerRef} name={"droppable"} {...provided.droppableProps}>
```
```suggestion
                <Elem ref={provided.innerRef} name="droppable" {...provided.droppableProps}>
```

hlomzik commented on web/libs/datamanager/src/components/DataManager/DataManager.jsx:
```diff
@@ -66,32 +67,56 @@ const TabsSwitch = switchInjector(
+                    tab.setTitle(title);
+                    tab.save();
+                  }}
+                  onDuplicate={() => views.duplicateView(tab)}
```
so it works like this? even if the `tab` is potentially removed from the tree?
what about closing tab which was moved? and can we also apply this change to `onClose()`? so it will be `views.deleteView(tab)`

juliosgarbi replied:
The tab is always on the tree when we try to duplicate it, even if we move the position and reorder the array, the problem that we were having is that for some reason we were loosing the `tab.parent` reference, and because of this, we couldn't get the duplicateView method from views. Yeah make totally sense, I already applied the same to onClose(), thanks @hlomzik

juliosgarbi committed:
apply view to onClose

juliosgarbi commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10061829249)
> Successfully merged:  29 files changed, 302 insertions(+), 96 deletions(-)

juliosgarbi committed:
Merge branch 'develop' into 'fb-leap-1199'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10061829249

RiccardoMaistri commented:
Good morning!
Does the drag and drop funcitonality has to be enabled or enteprse version is required? I am running Label studio v 1.13.1 deployed using helm chart. I am using Version 130.0.6723.91 (Official Build) (64-bit)
